mock = {
    "48c1a1bc-31bd-4420-9d28-9e4e6289eb16": {
        "description": (
            "Significant heterogeneities in gene expression among "
            "individual cells are typically interrogated using single "
            "whole cell approaches. However, tissues that have highly "
            "interconnected processes, such as in the brain, present "
            "unique challenges. Single-nucleus RNA sequencing (SNS) has "
            "emerged as an alternative method of assessing a cell’s "
            "transcriptome through the use of isolated nuclei. However, "
            "studies directly comparing expression data between nuclei "
            "and whole cells are lacking. Here, we have characterized nuclear "
            "and whole cell transcriptomes in mouse single neurons and "
            "provided a normalization strategy to reduce method-specific "
            "differences related to the length of genic regions. We "
            "confirmed a high concordance between nuclear and whole "
            "cell transcriptomes in the expression of cell type and "
            "metabolic modeling markers, but less so for a subset of "
            "genes associated with mitochondrial respiration. Therefore, "
            "our results indicate that single-nucleus transcriptome sequencing "
            "provides an effective means to profile cell type expression "
            "dynamics in previously inaccessible tissues."
        ),
        "cell_count": 0,
    },
    "4e6f083b-5b9a-4393-9890-2a83da8188f1": {
        "description": (
            "As organisms age, cells accumulate genetic and "
            "epigenetic changes that eventually lead to "
            "impaired organ function or catastrophic failure "
            "such as cancer. Here we describe a single-cell "
            "transcriptome analysis of 2544 human pancreas "
            "cells from donors, spanning six decades of life. "
            "We find that islet cells from older donors have "
            "increased levels of disorder as measured both "
            "by noise in the transcriptome and by the number "
            "of cells which display inappropriate hormone "
            "expression, revealing a transcriptional instability "
            "associated with aging. By analyzing the spectrum "
            "of somatic mutations in single cells from "
            "previously-healthy donors, we find a specific "
            "age-dependent mutational signature characterized "
            "by C to A and C to G transversions, indicators "
            "of oxidative stress, which is absent in single "
            "cells from human brain tissue or in a tumor cell "
            "line. Cells carrying a high load of such mutations "
            "also express higher levels of stress and senescence "
            "markers, including FOS, JUN, and the cytoplasmic "
            "superoxide dismutase SOD1, markers previously linked "
            "to pancreatic diseases with substantial age-dependent "
            "risk, such as type 2 diabetes mellitus and adenocarcinoma. "
            "Thus, our single-cell approach unveils gene expression "
            "changes and somatic mutations acquired in aging human "
            "tissue, and identifies molecular pathways induced by "
            "these genetic changes that could influence human "
            "disease. Also, our results demonstrate the feasibility "
            "of using single-cell RNA-seq data from primary "
            "cells to derive meaningful insights into the "
            "genetic processes that operate on aging human "
            "tissue and to determine which molecular mechanisms "
            "are coordinated with these processes. Examination "
            "of single cells from primary human pancreas tissue"
        ),
        "cell_count": 2544,
    },
    "c4077b3c-5c98-4d26-a614-246d12c2e5d7": {
        "description": (
            "Assessment of the effect of cold ischaemic time on single cell RNAseq data from human tissues using 10x "
            "Genomics 3' single cell RNA sequencing. This project contains data for spleen, oesophagus epithelium and "
            "lung parenchyma, three tissues that had previously been reported to have differential sensitivity to "
            "ischaemia. Samples were collected into Hypothermasol FRS hypothermic preservation media and dissociated "
            "fresh (as soon as possible) or at 12h, 24h, 72h post onset of cold ischaemia in the donor. Single cell "
            "and bulk RNA sequencing data was generated at each time point and whole genome sequencing was carried "
            "out for each donor. "
        ),
        "cell_count": 200000,
    },
    "4a95101c-9ffc-4f30-a809-f04518a23803": {
        "description": (
            "Human T cells coordinate adaptive immunity in diverse anatomic compartments through production of "
            "cytokines and effector molecules, but it is unclear how tissue site influences T cell persistence and "
            "function. Here, we use single cell RNA-sequencing (scRNA-seq) to define the heterogeneity of human T "
            "cells isolated from lungs, lymph nodes, bone marrow and blood, and their functional responses following "
            "stimulation. Through analysis of >50,000 resting and activated T cells, we reveal tissue T cell "
            "signatures in mucosal and lymphoid sites, and lineage-specific activation states across all sites "
            "including distinct effector states for CD8+ T cells and an interferon-response state for CD4+ T cells. "
            "Comparing scRNA-seq profiles of tumor-associated T cells to our dataset reveals predominant activated "
            "CD8+ compared to CD4+ T cell states within multiple tumor types. Our results therefore establish a high "
            "dimensional reference map of human T cell activation in health for analyzing T cells in disease. "
        ),
        "cell_count": 0,
    },
    "88ec040b-8705-4f77-8f41-f81e57632f7d": {
        "description": (
            "A multitude of single-cell RNA sequencing methods have been developed in recent years, with dramatic "
            "advances in scale and power, and enabling major discoveries and large scale cell mapping efforts. "
            "However, these methods have not been systematically and comprehensively benchmarked. Here, we directly "
            "compare seven methods for single cell and/or single nucleus profiling from three types of samples – cell "
            "lines, peripheral blood mononuclear cells and brain tissue – generating 36 libraries in six separate "
            "experiments in a single center. To analyze these datasets, we developed and applied scumi, a flexible "
            "computational pipeline that can be used for any scRNA-seq method. We evaluated the methods for both "
            "basic performance and for their ability to recover known biological information in the samples. Our "
            "study will help guide experiments with the methods in this study as well as serve as a benchmark for "
            "future studies and for computational algorithm development. "
        ),
        "cell_count": 59800,
    },
    "a9c022b4-c771-4468-b769-cabcf9738de3": {
        "description": (
            "The Human Cell Atlas project seeks to identify and functionally characterize all cell types in the human "
            "body. Single-cell RNA-sequencing (scRNA-seq) methods quantify gene expression in individual cells. In "
            "principle, highly parallelized scRNA-seq approaches enable marker-free decomposition of cells from "
            "complex tissues to identify known and unknown cell types and provide direct insights into regulatory and "
            "functional states of each cell. Different scRNA-seq technologies vary in cell capture efficiency, "
            "library preparation, and throughput. Importantly, these approaches also vary in sensitivity and accuracy "
            "of mRNA quantification. The Human Cell Atlas project will collate data generated across many labs, "
            "necessitating the evaluation of experimental replicability, and comparing different experimental "
            "procedures. It is important to develop robust sequencing protocols and benchmark their performance on "
            "select cells and tissues. This project will perform a systematic comparison of three single cell RNA "
            "sequencing technologies, viz., Drop-Seq, Fluidigm C1, and DroNC-Seq and resulting datasets. "
            "Droplet-based methods facilitate high-throughput processing of tens of thousands of cells in parallel "
            "and at low costs, making them ideal for the Human Cell Atlas project. This project proposes to compare "
            "the sensitivity, accuracy, and precision with which the three methods quantify transcript levels within "
            "individual cells. In addition, this project will directly compare these methods using a ‘synthetic "
            "tissue’ created from a mixture of multiple cell types at known ratios. "
        ),
        "cell_count": 0,
    },
    "e0009214-c0a0-4a7b-96e2-d6a83e966ce0": {
        "description": (
            "We have created a compendium of single cell transcriptome data from the model organism Mus musculus "
            "comprising more than 100,000 cells from 20 organs and tissues. These data represent a new resource for "
            "cell biology, revealing gene expression in poorly characterized cell populations and allowing for direct "
            "and controlled comparison of gene expression in cell types shared between tissues, such as T-lymphocytes "
            "and endothelial cells from distinct anatomical locations. Two distinct technical approaches were used "
            "for most tissues: one approach, microfluidic droplet-based 3’-end counting, enabled the survey of "
            "thousands of cells at relatively low coverage, while the other, FACS-based full length transcript "
            "analysis, enabled characterization of cell types with high sensitivity and coverage. The cumulative data "
            "provide the foundation for an atlas of transcriptomic cell biology. "
        ),
        "cell_count": 49000,
    },
    "cc95ff89-2e68-4a08-a234-480eca21ce79": {
        "description": (
            "Diverse cells of the immune system maintain and protect tissue function, integrity, and homeostasis upon "
            "changes in functional demands and diverse perturbations. Recent advances such as massively parallel "
            "single-cell RNA-sequencing and sophisticated computational methods help shed new light on this "
            "complexity. This immune cell census aims to profile up to 2M immunocytes, the first tranche of this is "
            "currently available. With computational methods optimized to a massive scale, we can readily identify "
            "cell types and markers, as well as the process of hematopoietic differentiation. The high quality and "
            "comprehensive reference map is provided as an open community resource for understanding human health and "
            "disease. "
        ),
        "cell_count": 528100,
    },
    "008e40e8-66ae-43bb-951c-c073a2fa6774": {
        "description": (
            "This project is part of the Swedish Human Developmental Cell Atlas: a comprehensive molecular atlas of "
            "human prenatal development, describing every cell type in molecular detail, and showing the spatial "
            "distribution of cells in three dimensions and their development over time. The project will focus on "
            "developmental tissue from brain, lung and heart. This part of the project is generating single cell gene "
            "expression profiles using high-throughput single cell mRNA sequencing with the aim to identify and "
            "characterize molecularly defined celltypes. "
        ),
        "cell_count": 0,
    },
    "a29952d9-925e-40f4-8a1c-274f118f1f51": {
        "description": (
            "Our aims are to generate a full representation of human hematopoiesis in blood and bone marrow of humans "
            "using a multi-tier and iterative collection and analysis of 200,000 cells from ten healthy human bone "
        ),
        "cell_count": 9500,
    },
    "ae71be1d-ddd8-4feb-9bed-24c3ddb6e1ad": {
        "description": (
            "We used single-cell RNA-sequencing to generate transcriptional profiles of endocrine and exocrine cell "
            "types of the human pancreas. Pancreatic tissue and islets were obtained from six healthy and four T2D "
            "cadaveric donors. Islets were cultured and dissociated into single-cell suspension. Viable individual "
            "cells were distributed via fluorescence-activated cell sorted (FACS) into 384-well plates containing "
            "lysis buffer. Single-cell cDNA libraries were generated using the Smart-seq2 protocol. Gene expression "
            "was quantified as reads per kilobase transcript and per million mapped reads (RPKM) using rpkmforgenes. "
            "Bioinformatics analysis was used to classify cells into cell types without knowledge of cell types or "
            "prior purification of cell populations. We revealed subpopulations in endocrine and exocrine cell types, "
            "identified genes with interesting correlations to body mass index (BMI) in specific cell types and found "
            "transcriptional alterations in T2D. Complementary whole-islet RNA-seq data have also been deposited at "
            "ArrayExpress under accession number E-MTAB-5060 ("
            "http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5060). "
        ),
        "cell_count": 3500,
    },
    "091cf39b-01bc-42e5-9437-f419a66c8a45": {
        "description": (
            "Differentiation is among the most fundamental processes in cell biology. Single cell RNA-seq studies "
            "have demonstrated that differentiation is a continuous process and in particular cell states are "
            "observed to reside on largely continuous spaces. We have developed Palantir, a graph based algorithm to "
            "model continuities in cell state transitions and cell fate choices. Modeling differentiation as a Markov "
            "chain, Palantir determines probabilities of reaching terminal states from cells in each intermediate "
            "state. The entropy of these probabilities represent the differentiation potential of the cell in the "
            "corresponding state. Applied to single cell RNA-seq dataset of CD34+ hematopoietic cells from human bone "
            "marrows, Palantir accurately identified key events leading up to cell fate commitment. Integration with "
            "ATAC-seq data from bulk sorted populations helped identify key regulators that correlate with cell fate "
            "specification and commitment. "
        ),
        "cell_count": 1500000,
    },
    "f83165c5-e2ea-4d15-a5cf-33f3550bffde": {
        "description": (
            "During early human pregnancy the uterine mucosa transforms into the decidua, into which the fetal "
            "placenta implants and where placental trophoblast cells intermingle and communicate with maternal cells. "
            "Trophoblast–decidual interactions underlie common diseases of pregnancy, including pre-eclampsia and "
            "stillbirth. Here we profile the transcriptomes of about 70,000 single cells from first-trimester "
            "placentas with matched maternal blood and decidual cells. The cellular composition of human decidua "
            "reveals subsets of perivascular and stromal cells that are located in distinct decidual layers. There "
            "are three major subsets of decidual natural killer cells that have distinctive immunomodulatory and "
            "chemokine profiles. We develop a repository of ligand–receptor complexes and a statistical tool to "
            "predict the cell-type specificity of cell–cell communication via these molecular interactions. Our data "
            "identify many regulatory interactions that prevent harmful innate or adaptive immune responses in this "
            "environment. Our single-cell atlas of the maternal–fetal interface reveals the cellular organization of "
            "the decidua and placenta, and the interactions that are critical for placentation and reproductive "
            "success. "
        ),
        "cell_count": 0,
    },
    "90bd6933-40c0-48d4-8d76-778c103bf545": {
        "description": (
            "CD4+ cytotoxic T lymphocytes (CD4-CTLs) have been reported to play a protective role in several viral "
            "infections. However, little is known in humans about the biology of CD4-CTL generation, their functional "
            "properties, heterogeneity and clonal diversity, especially in relation to other well-described CD4+ "
            "memory T cell subsets. We performed single-cell RNA-seq in over 9000 cells to unravel CD4-CTL "
            "heterogeneity, transcriptional profile and clonality in humans. The single-cell differential gene "
            "expression analysis, revealed a spectrum of known transcripts, including several linked to cytotoxic and "
            "co-stimulatory function, and transcripts of unknown cytotoxicity-related function that are expressed at "
            "higher levels in the TEMRA subset, which is highly enriched for CD4-CTLs, compared to cells in the "
            "central and effector memory subsets (TCM, TEM). Simultaneous T cells antigen receptor (TCR) analysis in "
            "single-cells and bulk subsets revealed that CD4-TEMRA cells show marked clonal expansion compared to TCM "
            "and TEM cells and that the majority of CD4-TEMRA were dengue virus (DENV)-specific in subjects with "
            "previous DENV infection. The profile of CD4-TEMRA was highly heterogeneous across subjects, "
            "with four distinct clusters identified by the single-cell analysis. Most importantly, we identified "
            "distinct clusters of CD4-CTL effector and precursor cells in the TEMRA subset; the precursor cells "
            "shared TCR clonotypes with CD4-CTL effectors and were distinguished by high expression of the "
            "interleukin-7 receptor. Our identification of a CD4-CTL precursor population may allow further "
            "investigation of how CD4-CTLs arise in humans and thus could provide insights into the mechanisms that "
            "may be utilized to generate durable and effective CD4-CTL immunity. "
        ),
        "cell_count": 2200,
    },
    "005d611a-14d5-4fbf-846e-571a1f874f70": {
        "description": (
            "The purpose of this project is to assess the relevance of pluripotent stem cell-derived cerebral and "
            "liver organoids to recapitulate the variation in cell-type specific gene expression programs between "
            "individuals. Towards this aim, we will generate reference atlases of the developing cortex and liver "
            "from multiple individuals, derive iPSC lines from these same individuals, and determine if "
            "inter-individual gene expression variation is recapitulated in cerebral and liver organoids from the "
            "same individual from which we have reference maps. In parallel we will assess the genetic contribution "
            "to variablity between organoids from different iPSCs of multiple human individuals that are available in "
            "existing iPSC resources (e.g. HipSci). "
        ),
        "cell_count": 19900,
    },
    "f81efc03-9f56-4354-aabb-6ce819c3d414": {
        "description": (
            "Despite rapid developments in single cell sequencing technology, sample-specific batch effects, "
            "detection of cell doublets, and the cost of generating massive datasets remain outstanding challenges. "
            "Here, we introduce cell hashing, where oligo-tagged antibodies against ubiquitously expressed surface "
            "proteins are used to uniquely label cells from distinct samples, which can be subsequently pooled. By "
            "sequencing these tags alongside the cellular transcriptome, we can assign each cell to its sample of "
            "origin, and robustly identify doublets originating from multiple samples. We demonstrate our approach by "
            "pooling eight human PBMC samples on a single run of the 10x Chromium system, substantially reducing our "
            "per-cell costs for library generation. Cell hashing is inspired by, and complementary to, "
            "elegant multiplexing strategies based on genetic variation, which we also leverage to validate our "
            "results. We therefore envision that our approach will help to generalize the benefits of single cell "
            "multiplexing to diverse samples and experimental designs "
        ),
        "cell_count": 20000,
    },
    "abe1a013-af7a-45ed-8c26-f3793c24a1f4": {
        "description": (
            "Tissue-resident immune cells are important for organ homeostasis and defense. The epithelium may "
            "contribute to these functions directly, or via crosstalk with immune cells. We resolved the "
            "spatio-temporal immune topology of the human kidney with a scRNAseq dataset from 40,268 mature and 27,"
            "203 fetal kidney cells. Within the epithelial compartment there were zonated patterns of immune gene "
            "expression, with pelvic epithelial anti-microbial peptide expression in mature but not fetal kidneys. "
            "Tissue-resident immune cells were evident in both fetal and mature kidney, with post-natal acquisition "
            "of infection-defense capabilities. Epithelial-immune signalling orchestrated localization of macrophages "
            "and neutrophils to regions of the kidney susceptible to infection. "
        ),
        "cell_count": 0,
    },
    "f8aa201c-4ff1-45a4-890e-840d63459ca2": {
        "description": (
            "Intestinal mesenchymal cells play essential roles in epithelial homeostasis, matrix remodeling, "
            "immunity, and inflammation. But the extent of heterogeneity within the colonic mesenchyme in these "
            "processes remains unknown. Using unbiased single-cell profiling of over 16,500 colonic mesenchymal "
            "cells, we reveal four subsets of fibroblasts expressing divergent transcriptional regulators and "
            "functional pathways, in addition to pericytes and myofibroblasts. We identified a niche population "
            "located in proximity to epithelial crypts expressing SOX6, F3 (CD142), and WNT genes essential for "
            "colonic epithelial stem cell function. In colitis, we observed dysregulation of this niche and emergence "
            "of an activated mesenchymal population. This subset expressed TNF superfamily member 14 (TNFSF14), "
            "fibroblastic reticular cell-associated genes, IL-33, and Lysyl oxidases. Further, it induced factors "
            "that impaired epithelial proliferation and maturation and contributed to oxidative stress and disease "
            "severity in vivo. Our work defines how the colonic mesenchyme remodels to fuel inflammation and barrier "
            "dysfunction in IBD. "
        ),
        "cell_count": 16900,
    },
    "116965f3-f094-4769-9d28-ae675c1b569c": {
        "description": (
            "Dendritic cells (DCs) are professional antigen-presenting cells specialized in the recognition, "
            "processing and presentation of antigens to T-cells, inducing adaptive immune responses. We have "
            "identified PU.1, IRF8 and BATF3 transcription factors (TFs) to reprogram human fibroblasts into DC type "
            "1 (DC1). Induced DCs acquire a step-wise DC1 transcriptional signature at the single cell level as "
            "assessed with droplet based scRNA-seq. Hence, we provide evidence that antigen presentation can be "
            "dynamically programmed by a small combination of TFs in human cells. Moreover, it represents a platform "
            "for generating autologous DC1, the most attractive DC subset for cancer immunotherapy. "
        ),
        "cell_count": 0,
    },
    "f86f1ab4-1fbb-4510-ae35-3ffd752d4dfc": {
        "description": (
            "Although the function of the mammalian pancreas hinges on complex interactions of distinct cell types, "
            "gene expression profiles have primarily been described with bulk mixtures. Here we implemented a "
            "droplet-based, single-cell RNA-seq method to determine the transcriptomes of over 12,000 individual "
            "pancreatic cells from four human donors and two mouse strains. Cells could be divided into 15 clusters "
            "that matched previously characterized cell types: all endocrine cell types, including rare "
            "epsilon-cells; exocrine cell types; vascular cells; Schwann cells; quiescent and activated stellate "
            "cells; and four types of immune cells. We detected subpopulations of ductal cells with distinct "
            "expression profiles and validated their existence with immuno-histochemistry stains. Moreover, "
            "among human beta- cells, we detected heterogeneity in the regulation of genes relating to functional "
            "maturation and levels of ER stress. Finally, we deconvolved bulk gene expression samples using the "
            "single-cell data to detect disease-associated differential expression. Our dataset provides a resource "
            "for the discovery of novel cell type-specific transcription factors, signaling receptors, and medically "
            "relevant genes. "
        ),
        "cell_count": 12400,
    },
    "027c51c6-0719-469f-a7f5-640fe57cbece": {
        "description": (
            "Single cell genomics are revolutionizing our ability to characterize complex tissues. By contrast, "
            "the techniques used to analyze the renal biopsy are little changed over the past several decades. In "
            "this study we have tested the hypothesis that single cell RNA sequencing (sc-RNA-seq) can "
            "comprehensively describe cell types and states in a human kidney biopsy. We generated 8,746 single cell "
            "cell transcriptomes from a healthy adult kidney and a single kidney transplant biopsy core by "
            "sc-RNA-seq. Unsupervised clustering analysis of the biopsy identified 16 distinct cell types, "
            "including all of the major immune cell types and most native kidney cell types in this biopsy whose "
            "histologic read was mixed rejection. Endothelial cells formed three distinct sub-clusters: resting cells "
            "and two activated endothelial cell clusters, including expression of Fc-receptor pathway activation and "
            "immunoglobulin internalization genes, consistent with the pathologic diagnosis of antibody-mediated "
            "rejection. Monocytes formed two sub-clusters representing a non-classical CD16+ group and a classical "
            "CD16- group expressing dendritic cell maturation markers. The presence of both monocyte cell subtypes "
            "was validated by staining of of independent transplant biopsies. Comparison of healthy kidney epithelial "
            "transcriptomes to their biopsy counterparts identified novel segment specific pro-inflammatory "
            "responses. We mapped previously defined genes that associate with rejection outcomes to single cell "
            "types and generated a searchable online gene expression database. These findings represent a first step "
            "towards incorporation of single cell transcriptomics into kidney biopsy interpretation, describe a "
            "heterogeneous immune response in mixed rejection and provide a searchable resource for the scientific "
            "community. "
        ),
        "cell_count": 4000,
    },
    "577c946d-6de5-4b55-a854-cd3fde40bff2": {
        "description": (
            "Diabetic nephropathy is characterized by damage to both the glomerulus and tubulointerstitium, "
            "but relatively little is known about accompanying cell-specific changes in gene expression. We performed "
            "unbiased single nucleus RNA sequencing (snRNAseq) on cryopreserved human diabetic kidney samples to "
            "generate 23,980 single nucleus transcriptomes from three control and three early diabetic nephropathy "
            "samples. All major cell types of the kidney were represented in the final dataset. Side by side "
            "comparison demonstrated cell-type-specific changes in gene expression that are important for ion "
            "transport, angiogenesis, and immune cell activation. In particular, we show that the diabetic thick "
            "ascending limb, late distal convoluted tubule, and principal cells all adopt a gene expression signature "
            "consistent with increased potassium secretion, including alterations in Na-K+-ATPase, WNK1, "
            "mineralocorticoid receptor and NEDD4L expression, as well as decreased paracellular calcium and "
            "magnesium reabsorption. We also identify strong angiogenic signatures in glomerular cell types, "
            "proximal convoluted tubule, distal convoluted tubule and principal cells. Taken together, these results "
            "suggest that increased potassium secretion and angiogenic signaling represent early kidney responses in "
            "human diabetic nephropathy. "
        ),
        "cell_count": 30000,
    },
    "2043c65a-1cf8-4828-a656-9e247d4e64f1": {
        "description": (
            "Diverse cell types are produced from dorsal and ventral regions of the developing neural tube. In this "
            "study we describe a system for generating human inhibitory interneurons by ventralizing human embryonic "
            "stem cells in vitro and characterizing the gene expression of the cell types produced over time. We "
            "engineered a DCX-Citrine/Y hESC line to sort and characterize progenitor and neuron transcriptomics "
            "separately at both the subpopulation and single cell level. The cells generated in vitro were compared "
            "to similar populations present in human fetal brain samples by mapping gene expression data from human "
            "fetal cells onto the principal component analysis (PCA) space of in vitro-derived populations. Weighted "
            "gene co-expression network analysis (WGCNA) was used to determine the discreet cell types present at "
            "D24, D54, D100 and D125 of culture, and describe the gene expression changes that occur in progenitor "
            "and neuron populations over time. Immature lateral ganglionic eminence and medial ganglionic eminence "
            "cells are present at early timepoints, along with MGE-like and dorsal pallium-like neuronal progenitors. "
            "At later timepoints we observe the emergence of SST-expressing interneurons, as well as oligodendrocyte "
            "and astrocyte progenitors. We also identified genes that were upregulated in somatostatin-expressing "
            "interneurons as they mature. The transcriptomes of 1732 ventralized single cells were profiled by "
            "SmartSeq2 at different timepoints throughout a 125-day differentiation protocol that converted H1 human "
            "embryonic stem cells to a variety of ventrally-derived cell types. "
        ),
        "cell_count": 0,
    },
    "74b6d569-3b11-42ef-b6b1-a0454522b4a0": {
        "description": (
            "Cortex, hippocampus, and subventricular zone were purchased from BrainBits (C57EHCV). They were from 2 "
            "E18 C57BL/6 mice dissected on the same day, shipped overnight on ice, and stored at 4C until being "
            "prepared for scRNA-Seq. Brain tissues were dissociated following the Demonstrated Protocol for Mouse "
            "Embryonic Neural Tissue (https://support.10xgenomics.com/single-cell/sample-prep/doc/demonstrated"
            "-protocol-dissociation-of-mouse-embryonic-neural-tissue-for-single-cell-rna-sequencing). 69 scRNA-Seq "
            "libraries were made from first mouse brain 2 days after the dissection. Another 64 scRNA-Seq libraries "
            "were made from second mouse brain 6 days after the dissection. "
        ),
        "cell_count": 1300000,
    },
    "8185730f-4113-40d3-9cc3-929271784c2b": {
        "description": (
            "The retina is a specialized neural tissue that senses light and initiates image processing. Although the "
            "functional organization of specific retina cells has been well studied, the molecular profile of many "
            "cell types remains unclear in humans. To comprehensively profile the human retina, we performed "
            "single‐cell RNA sequencing on 20,009 cells from three donors and compiled a reference transcriptome "
            "atlas. Using unsupervised clustering analysis, we identified 18 transcriptionally distinct cell "
            "populations representing all known neural retinal cells: rod photoreceptors, cone photoreceptors, "
            "Müller glia, bipolar cells, amacrine cells, retinal ganglion cells, horizontal cells, astrocytes, "
            "and microglia. Our data captured molecular profiles for healthy and putative early degenerating rod "
            "photoreceptors, and revealed the loss of MALAT1 expression with longer post‐mortem time, "
            "which potentially suggested a novel role of MALAT1 in rod photoreceptor degeneration. We have "
            "demonstrated the use of this retina transcriptome atlas to benchmark pluripotent stem cell‐derived cone "
            "photoreceptors and an adult Müller glia cell line. This work provides an important reference with "
            "unprecedented insights into the transcriptional landscape of human retinal cells, which is fundamental "
            "to understanding retinal biology and disease. "
        ),
        "cell_count": 44000,
    },
    "8c3c290d-dfff-4553-8868-54ce45f4ba7f": {
        "description": (
            "The cancer microenvironment is a complex ecosystem characterized by dynamic interactions between diverse "
            "cell types, including malignant, immune and stromal cells. Here, we performed single-cell RNA sequencing "
            "on CD45+ and CD45- cells isolated from tumour and lymph nodes during a mouse model of melanoma. The "
            "transcriptional profiles of these individual cells taken at different time points coupled with assembled "
            "T cell receptor sequences, allowed us to identify distinct immune subpopulations and delineate their "
            "developmental trajectory. Our study provides insights into the complex interplay among cells within the "
            "tumour microenvironment and presents a valuable resource for future translational applications. "
        ),
        "cell_count": 6600,
    },
    "4d6f6c96-2a83-43d8-8fe1-0f53bffd4674": {
        "description": (
            "The liver is the largest solid organ in the body and is critical for metabolic and immune functions. "
            "However, little is known about the cells that make up the human liver and its immune microenvironment. "
            "Here we report a map of the cellular landscape of the human liver using single-cell RNA sequencing. We "
            "provide the transcriptional profiles of 8444 parenchymal and non-parenchymal cells obtained from the "
            "fractionation of fresh hepatic tissue from five human livers. Using gene expression patterns, "
            "flow cytometry, and immunohistochemical examinations, we identify 20 discrete cell populations of "
            "hepatocytes, endothelial cells, cholangiocytes, hepatic stellate cells, B cells, conventional and "
            "non-conventional T cells, NK-like cells, and distinct intrahepatic monocyte/macrophage populations. "
            "Together, our study presents a comprehensive view of the human liver at single-cell resolution that "
            "outlines the characteristics of resident cells in the liver, and in particular provides a map of the "
            "human hepatic immune microenvironment. "
        ),
        "cell_count": 30000,
    },
    "1defdada-a365-44ad-9b29-443b06bd11d6": {
        "description": (
            "Across the animal kingdom, gastrulation represents a key developmental event during which embryonic "
            "pluripotent cells diversify into lineage-specific precursors that will generate the adult organism. Here "
            "we report the transcriptional profiles of 116,312 single cells from mouse embryos collected at nine "
            "sequential time points ranging from 6.5 to 8.5 days post-fertilization. We construct a molecular map of "
            "cellular differentiation from pluripotency towards all major embryonic lineages, and explore the complex "
            "events involved in the convergence of visceral and primitive streak-derived endoderm. "
        ),
        "cell_count": 339100,
    },
    "cddab57b-6868-4be4-806f-395ed9dd635a": {
        "description": (
            "As organisms age, cells accumulate genetic and epigenetic changes that eventually lead to impaired organ "
            "function or catastrophic failure such as cancer. Here we describe a single-cell transcriptome analysis "
            "of 2544 human pancreas cells from donors, spanning six decades of life. We find that islet cells from "
            "older donors have increased levels of disorder as measured both by noise in the transcriptome and by the "
            "number of cells which display inappropriate hormone expression, revealing a transcriptional instability "
            "associated with aging. By analyzing the spectrum of somatic mutations in single cells from "
            "previously-healthy donors, we find a specific age-dependent mutational signature characterized by C to A "
            "and C to G transversions, indicators of oxidative stress, which is absent in single cells from human "
            "brain tissue or in a tumor cell line. Cells carrying a high load of such mutations also express higher "
            "levels of stress and senescence markers, including FOS, JUN, and the cytoplasmic superoxide dismutase "
            "SOD1, markers previously linked to pancreatic diseases with substantial age-dependent risk, such as type "
            "2 diabetes mellitus and adenocarcinoma. Thus, our single-cell approach unveils gene expression changes "
            "and somatic mutations acquired in aging human tissue, and identifies molecular pathways induced by these "
            "genetic changes that could influence human disease. Also, our results demonstrate the feasibility of "
            "using single-cell RNA-seq data from primary cells to derive meaningful insights into the genetic "
            "processes that operate on aging human tissue and to determine which molecular mechanisms are coordinated "
            "with these processes. Examination of single cells from primary human pancreas tissue "
        ),
        "cell_count": 38160,
    },
    "9c20a245-f2c0-43ae-82c9-2232ec6b594f": {
        "description": (
            "The human neural retina is a heterogenous tissue that is an ordered array of approximately 70 different "
            "types of neurons, each playing a unique role in processing the visual signal. Currently, "
            "the transcriptomic-level classification of human retinal cell types is still limited. Using "
            "single-nuclei RNA-seq, a comprehensive single-cell profiling study has been carried out for healthy "
            "human retina from four individual donors. With the transcriptomic profiles for a total of over 200K "
            "nuclei being generated, we aim at revealing the transcriptome profiles for most subtypes of the retinal "
            "cells. Comparison with previous published macaque and mice dataset indicates an overall good alignment "
            "among species. In addition, the gene expression profile that is unique to human has been discovered. "
            "Furthermore, over 200 genes associated with human retinal diseases are investigated in our dataset and "
            "are found to show both cell type and regional specificity. In summary, results from our study serves as "
            "not only a cell atlas reference but also the foundation for future studies of human retinal diseases. "
        ),
        "cell_count": 230800,
    },
}
